docs/architecture/adr-0026-post-hoc-metrics-scripting.md

```diff
@@ -0,0 +1,155 @@
+to the database or API as metrics.
+
+Choosing this strategy will require:
+
+- an API endpoint to POST metrics with superuser auth only (unless we can ship the Dao with
+  Lambda functions)
+- A Lambda function to kick off an Athena query for each metric we want to calculate
+- A Lambda function to check whether an Athena query (by ID) is complete
+- A Step Function to organize the function executions
+- A Lambda function to parse Athena query results into Metric events and ship them to the
+  API
+
+### Non-events
+
+For things that aren't events, we have a choice between extending the database schema
+to support querying values we care about (e.g., add a "file size" attribute to `Upload`s)
+and calculating usage from objects in S3 (e.g., check the combined sizes of a bunch of
+keys, based on some database query).
+
+#### Option 1: Extending the database
+
+Extensions we know would be necessary are adding `ingest_size_bytes` back to the scenes
+table and adding `file_size` to `uploads`.  Adding these columns would be just a migration.
+However, we'd need a one-off asynchronous process to backfill the file size information for
+existing scenes and uploads. Additionally, since uploads and scenes can come from anywhere,
+we'd need to add special handling to check and fill in upload file size and scene ingest size
+only when the files / scenes are stored in Raster Foundry S3 buckets. Finally, for
+consistency, we should also delete objects in S3 on scene and upload deletion, if we own the
+bucket location, since otherwise our reported usage numbers will underestimate actual usage.
+Finally, we should  update database records when we cleanup unused objects, since if we don't,
+we'll overestimate actual usage.
+
+Complexity in this example in general depends on trying to maintain consistent state between
+the database and what lives in S3. This is going to be hard, and we're probably going to miss
+sometimes. That's especially costly if we're measuring usage for billing purposes.
+
+#### Option 2: Just get the information from s3
+
+We have a few options here, but the general theme here is "S3 knows how big things are, so
+let it tell us about it". We've been consistent enough in where ingests and uploads go that
+it won't be difficult for us to tell where to look for things or who owns them. The workflow
+in this case can either be reactive, with Lambda functions responding to S3 events based on
+a bucket filter and updating database records accordingly, or periodic and the result of some scheduled roll-up operation.
+
+In either case, these events need to go somewhere. Since in this case we're not extending the
+database to include this information on the objects themselves, we should keep a separate table
+of usage information. This table could look something like:
+
+```sql
+CREATE TYPE metered_object_type AS ENUM ('SCENE', 'UPLOAD');
+
+CREATE TABLE usage_data (
+  metered_object_type metered_object_type NOT NULL,
+  object_id uuid NOT NULL,
```

Comment:
For uploads that have more than one file does this schema mean there would be multiple records for a single upload? I don't think that's a problem, but want to make sure we cover that case.

Reply:
I'm not sure -- I don't think that's necessary for tracking metrics this way, since we can just add file sizes together, but it might be clearer if we do start treating uploads that way